solana/src/lib.rs

```diff
@@ -0,0 +1,111 @@
+
+    pub fn stake_tokens(ctx: Context<StakeToken>,amount:u64)->Result<()>{
+
+        // Transfer tokens from the user's account to the pool's account
+        let cpi_accounts = Transfer {
+            from: ctx.accounts.user_token_account.to_account_info(),
+            to: ctx.accounts.pool_token_account.to_account_info(),
+            authority: ctx.accounts.user.to_account_info(),
+        };
+        let cpi_program = ctx.accounts.token_program.to_account_info();
+        let cpi_ctx = CpiContext::new(cpi_program, cpi_accounts);
+        token::transfer(cpi_ctx, amount)?;
+
+        // Mint equivalent amount of reward tokens to the user's reward token account
+        let mint_cpi_accounts = MintTo {
+            mint: ctx.accounts.reward_token_mint.to_account_info(),
+            to: ctx.accounts.user_reward_token_account.to_account_info(),
+            authority: ctx.accounts.mint_authority.to_account_info(),
+        };
+        let mint_cpi_program = ctx.accounts.token_program.to_account_info();
+        let mint_cpi_ctx = CpiContext::new(mint_cpi_program, mint_cpi_accounts);
+        token::mint_to(mint_cpi_ctx, amount)?;
+
+        Ok(())
+    }
+
+    pub fn unstake_tokens(ctx: Context<UnstakeTokens>, pool_seed: Vec<u8>,bump:u8, amount: u64) -> Result<()> {
+    // Burn the equivalent amount of reward tokens from the user's account
+        let burn_cpi_accounts = Burn {
+            mint: ctx.accounts.reward_token_mint.to_account_info(),
+            from: ctx.accounts.user_reward_token_account.to_account_info(),
+            authority: ctx.accounts.burn_authority.to_account_info(),
+        };
+        let burn_cpi_program = ctx.accounts.token_program.to_account_info();
+        let burn_cpi_ctx = CpiContext::new(burn_cpi_program, burn_cpi_accounts);
+        token::burn(burn_cpi_ctx, amount)?;
+
+        // Transfer the originally staked tokens from the pool's account back to the user's account
+        let seeds = &[pool_seed.as_slice(), &[bump]];
+        let signer = &[&seeds[..]];
+        let transfer_cpi_accounts = Transfer {
+            from: ctx.accounts.pool_token_account.to_account_info(),
+            to: ctx.accounts.user_token_account.to_account_info(),
+            authority: ctx.accounts.pool_token_account.to_account_info(), // This may need to be adjusted based on who has authority over the pool account
+        };
+        let transfer_cpi_program = ctx.accounts.token_program.to_account_info();
+        let transfer_cpi_ctx = CpiContext::new_with_signer(
+            transfer_cpi_program,
+            transfer_cpi_accounts,
+            signer,
+        );
+        token::transfer(transfer_cpi_ctx, amount)?;
+
+        Ok(())
+    }
+}
+
+#[derive(Accounts)]
+pub struct  StakeToken<'info> {
+    #[account(mut)]
+    pub user_token_account: Account<'info, TokenAccount>, // User's account for the staked tokens
+    #[account(init, payer = user, space = TokenAccount::LEN, seeds = [b"pool"], bump)]
```

Comment:
you would need to provide the token mint as well along with an authority which should be a PDA as well. 
```suggestion
    #[account(init, payer = user, seeds = [b"pool"], bump, token::mint = <TokenMint>, token::authority = <AnotherPDA>)]
```